File: list1/c/app/main.c
#include "../inc/lib.h"
#include <stdio.h>

static void diophantine_equation_test(void) {
  printf("================ [DIOPHANTINE  EQUASION] ================\n");
  enum { dimension  = 7 };
  uint64_t arr[dimension][3] = {{13, 5, 1}, {27, 4, 3}, {15, 5, 25}, {1, 1, 1}, {0, 0, 0}, {0, 0, 1}, {0, 1, 0}};

  for(int i = 0; i < dimension; i++) {
    int64_t x, y, a = arr[i][0], b = arr[i][1], c = arr[i][2];

    printf("\nDiophantine equation for a = %ld, b = %ld, c = %ld\n", a, b, c);
    printf("Iterative algorithm:\n");

    diophantine_equation(a, b, c, &x, &y);
    
    if (x != LLONG_MIN) {
      printf("Solution found!\n");
      printf("%lu * %ld + %lu * %ld = %lu\n\n", a, x, b, y, c);
    }

    printf("Recursive algorithm:\n");
    rec_diophantine_equation(a, b, c, &x, &y);

    if (x != LLONG_MIN) {
      printf("Solution found!\n");
      printf("%lu * %ld + %lu * %ld = %lu\n\n", a, x, b, y, c);
    }
    printf("---------------------------------------------------------\n");
  }
} 

int main(int argc, char const *argv[]) {
  (void)argc;
  (void)argv;

  uint64_t e = 6, f = 27, g = 51;

  printf("====================== [FACTORIAL] ======================\n\n");

  printf("Factorial:           %ld\n", factorial(e));
  printf("Recursive Facotrial: %ld\n\n", rec_factorial(e));

  printf("========================= [GCD] =========================\n\n");
  printf("GCD for           a = %ld, b = %ld: %ld\n", f, g, gcd(f, g));
  printf("Recursive GCD for a = %ld, b = %ld: %ld\n\n", f, g, rec_gcd(f, g));

  diophantine_equation_test();

  return 0;
}

File: list1/julia/main.jl
# julia main.jl
include("lib.jl")
using .LIB

TEST_ARRAY = [(13, 5, 1), (27, 4, 3), (15, 5, 25), (1, 1, 1), (0, 0, 0), (0, 0, 1), (0, 1, 0), (12, 73, 96)] 

function diophantine_equation_test() 
  println("================ [DIOPHANTINE  EQUASION] ================\n")
  for (a, b, c) in TEST_ARRAY
    println("Diophantine equation for a = $(a), b = $(b), c = $(c)\n")
    for (name, f) in [("Iterative ", gcd_extended), ("Recursive ", rec_gcd_extended)]
      println("$(name) algorithm:");
      (x, y) = diophantine_equation(a, b, c, f)
      if (x != typemin(Int64)) 
        println("Solution found!")
        println("$(a) * $(x) + $(b) * $(y) = $(c)\n");
      end
    end
    println("---------------------------------------------------------\n");
  end
end

function main(args::Array{String})
  x::UInt64 = 5
  y::UInt64 = 27
  z::UInt64 = 57

  println("====================== [FACTORIAL] ======================\n")
  println("Factorial:           $(factorial(x))")
  println("Recursive Facotrial: $(rec_factorial(x))\n")

  println("========================= [GCD] =========================\n")
  println("GCD for           a = $(y), b = $(z): $(gcd(y, z))")
  println("Recursive GCD for a = $(y), b = $(z): $(rec_gcd(y, z))\n")

  diophantine_equation_test()

end

if abspath(PROGRAM_FILE) == @__FILE__
  main(ARGS)
end
